docs: Define OpenTelemetry Component Inclusion Principles (#6773)
grafana/alloy: docs/developer/add-otel-component.md

Reviewer: Both this and the next file are about adding components to Alloy. Should we change the titles and filenames to make it clear that one is about adding component to OTel Engine and another is for adding component to Default Engine?

Author: I think for `adding-community-components` we can change the name to `adding-community-components-to-default-engine` to make that distinction clear.

`add-otel-component` addresses adding to _both_ engines though, and the inclusion criteria also apply to both engines, not just one or the other. I'm not sure if it makes sense to separate the contents out into two per-engine pages - do you feel it's too muddled to be in a single page?

Author: i've done this [here](https://github.com/grafana/alloy/pull/6773/commits/0c75263a9bd7f057a9f12fc43f384b832417a2a7)

### Signals against inclusion

These are the carrying costs and risks a component has to justify.

| Signal | Why it matters |
| --- | --- |
| Low stability or unmaintained | Contrib components vary widely in maturity, and an alpha or unmaintained component is a maintenance risk. |
| Non-reputable source repository | Components from unknown or inactive repositories carry higher trust and maintenance risk. |
| Heavy dependency or security footprint | Every bundled component adds ongoing cost, so a large dependency tree, a required fork, or known CVEs all raise the bar. |
| Incompatible license | Alloy is licensed under Apache 2.0, so a component must be under a compatible license to be bundled. |

### The decision process

These are principles, not a scoring system.
Maintainers weigh the signals together and make the final call, so meeting individual criteria doesn't guarantee inclusion.
A proposal results in one of three outcomes:

* **Bundled and maintained.** The component is added to both engines and maintained by the Alloy team.

> * **Community component.** The component is added to the Default Engine as a [community component](https://grafana.com/docs/alloy/latest/get-started/community_components/) with opt-in, best-effort support, and isn't bundled in the OTel Engine. You can still use it there through a custom OpenTelemetry Collector Builder (OCB) build. A community component still needs to serve a valid use case: even without a maintenance commitment, any heavy dependencies or security concerns it introduces remain the responsibility of the Alloy maintainers to ship.

Reviewer 2: ```suggestion
* **Community component.** The component is added to the Default Engine as a [community component](https://grafana.com/docs/alloy/latest/get-started/community_components/) with opt-in, best-effort support, and isn't bundled in the OTel Engine. You can still use it there through a custom OpenTelemetry Collector Builder (OCB) build. A community component still needs to serve a valid use case. Even without a maintenance commitment, the Alloy maintainers remain responsible for shipping any heavy dependencies or security concerns it introduces.
```

* **Not bundled.** The component isn't added to the distribution. You can still use it in your own build through OCB.

## Component lifecycle and removal

Bundled components follow the upstream OpenTelemetry Collector [component lifecycle](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md).

* If a component is marked **deprecated** upstream, it's also marked deprecated in the Alloy distribution and kept for at least two more minor releases before removal.
* If a component becomes **unmaintained** upstream, the same process applies and it's removed after it has been unmaintained for three months.

> In either case, notice is provided, and users can [create a custom build using OCB][ocb] to include a deprecated component.

Reviewer 2: ```suggestion
In either case, maintainers provide notice, and users can [create a custom build using OCB][ocb] to include a deprecated component.
```

We should say who provides notice.

This document uses a simple example to show how you add a component to Alloy that exists in an upstream OpenTelemetry Collector repository.
You can apply this information to add any part of an OpenTelemetry pipeline to Alloy.

## Inclusion criteria

The criteria below apply when a new component is proposed to be bundled into the official Alloy distribution.

Alloy ships a curated subset of upstream OpenTelemetry components rather than mirroring all of the OpenTelemetry Collector Contrib repository.
Keeping that set small is what lets the Alloy team maintain the bundled components and keep the dependency and security surface manageable.
These principles are about what the Alloy maintainers commit to maintain.

> Maintained components stay interoperable across both engines, so users can move between the Default Engine and the OTel Engine without losing functionality.

Reviewer: What does "interoperable" mean?

Author: As in, if a component exists in the default engine then it also exists in the otel engine - a user can migrate between the two

Reviewer: Do you mind rephrasing the sentence to avoid this term please? The way you said it here - "if a component exists in the default engine then it also exists in the otel engine" sounds more clear.

Author: done, there's some overlap with [this thread](https://github.com/grafana/alloy/pull/6773#discussion_r3703031145) so will resolve this to keep the discussion concise

> Components the Alloy maintainers don't intend to maintain become community components in the Default Engine and are excluded from the OTel Engine.

Reviewer: This makes it sound like a component becomes a community component when it's about to be deprecated. Components are usually added as community components when they first become available.

Author: hm, is it the last sentence that's making it seem that way? What about:

"Community components, which are outside our maintenance policy, live only in the default engine and are excluded from the OTel engine"

Reviewer: There is no need to say that community components aren't maintained by us. It's ok to say "Community components will not be added to OTel Engine. They will only be added to Default Engine.". If there's a link to the [community component page](https://grafana.com/docs/alloy/latest/get-started/components/community-components/) then that's enough.

Author: The thing is it ties directly into the point being made: we curate the list of components going into the engines because we want to be care about what we maintain - that ties into the stability/supportability of our distro. Leaving out the maintainability part kind of makes the mention of community components seem unrelated to the inclusion criteria

edit: I pushed some [changes](https://github.com/grafana/alloy/pull/6773/changes/34c15e187cac8e5a52d2bb575c14d9a90b3dae78) just to make things (imo) clearer, but i've kept the maintenance theme since it tied into the main point

## Add the component to the OTel engine

The steps above only register the component with Alloy's Default Engine. Alloy also ships an OTel Engine, an embedded OpenTelemetry Collector distribution run via `alloy otel`, with its own component set.

> Only non-community components are bundled in the OTel Engine. If your component is a [community component](https://grafana.com/docs/alloy/latest/get-started/community_components/), skip this step: community components live in the Default Engine only, and users who want them in the OTel Engine add them through a custom OCB build. For a non-community component, make it available in the OTel Engine by adding it to the OpenTelemetry Collector Builder (OCB) manifest at `collector/builder-config.yaml` under the matching category (`receivers`, `processors`, `exporters`, `extensions`, or `connectors`):

Reviewer: This could be phrased in more simple terms by just stating "OTel Engine does not support community components. If an OTel Engine user needs an extra component they can add it through OCB."

Author: yeah true - how's this? https://github.com/grafana/alloy/pull/6773/changes/a6ac35ba9ba5d51e76e1fd40757f5574c2eec4b9
